doxygen/man/man3/TransactionDesc.3
.TH "TransactionDesc" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TransactionDesc \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <transactiondesc\&.h>\fP
.PP
Inherits QObject\&.
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static QString \fBtoHTML\fP (\fBCWallet\fP *wallet, \fBCWalletTx\fP &wtx, \fBTransactionRecord\fP *rec, int unit)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Provide a human-readable extended HTML description of a transaction\&. 
.PP
Definition at line 13 of file transactiondesc\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "QString TransactionDesc::toHTML (\fBCWallet\fP * wallet, \fBCWalletTx\fP & wtx, \fBTransactionRecord\fP * rec, int unit)\fC [static]\fP"

.PP
Definition at line 41 of file transactiondesc\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/transactiondesc.cpp
#include "transactiondesc.h"

#include "darksilkunits.h"
#include "guiutil.h"

#include "base58.h"
#include "main.h"
#include "paymentserver.h"
#include "transactionrecord.h"
#include "timedata.h"
#include "ui_interface.h"
#include "wallet/wallet.h"
#include "txdb-leveldb.h"

#include <string>

QString TransactionDesc::FormatTxStatus(const CWalletTx& wtx)
{
    AssertLockHeld(cs_main);
    if (!IsFinalTx(wtx, nBestHeight + 1))
    {
        if (wtx.nLockTime < LOCKTIME_THRESHOLD)
            return tr("Open for %n more block(s)", "", wtx.nLockTime - nBestHeight);
        else
            return tr("Open until %1").arg(GUIUtil::dateTimeStr(wtx.nLockTime));
    }
    else
    {
        int nDepth = wtx.GetDepthInMainChain();
        if (nDepth < 0)
            return tr("conflicted");
        else if (GetAdjustedTime() - wtx.nTimeReceived > 2 * 60 && wtx.GetRequestCount() == 0)
            return tr("%1/offline").arg(nDepth);
        else if (nDepth < 10)
            return tr("%1/unconfirmed").arg(nDepth);
        else
            return tr("%1 confirmations").arg(nDepth);
    }
}

QString TransactionDesc::toHTML(CWallet *wallet, CWalletTx &wtx, TransactionRecord *rec, int unit)
{
    QString strHTML;

    LOCK2(cs_main, wallet->cs_wallet);
    strHTML.reserve(4000);
    strHTML += "<html><font face='verdana, arial, helvetica, sans-serif'>";

    int64_t nTime = wtx.GetTxTime();
    CAmount nCredit = wtx.GetCredit(ISMINE_ALL);
    CAmount nDebit = wtx.GetDebit(ISMINE_ALL);
    CAmount nNet = nCredit - nDebit;

    strHTML += "<b>" + tr("Status") + ":</b> " + FormatTxStatus(wtx);
    int nRequests = wtx.GetRequestCount();
    if (nRequests != -1)
    {
        if (nRequests == 0)
            strHTML += tr(", has not been successfully broadcast yet");
        else if (nRequests > 0)
            strHTML += tr(", broadcast through %n node(s)", "", nRequests);
    }
    strHTML += "<br>";

    strHTML += "<b>" + tr("Date") + ":</b> " + (nTime ? GUIUtil::dateTimeStr(nTime) : "") + "<br>";

    //
    // From
    //
    if (wtx.IsCoinBase() || wtx.IsCoinStake())
    {
        strHTML += "<b>" + tr("Source") + ":</b> " + tr("Generated") + "<br>";
    }
    else if (wtx.mapValue.count("from") && !wtx.mapValue["from"].empty())
    {
        // Online transaction
        strHTML += "<b>" + tr("From") + ":</b> " + GUIUtil::HtmlEscape(wtx.mapValue["from"]) + "<br>";
    }
    else
    {
        // Offline transaction
        if (nNet > 0)
        {
            // Credit
            if (CDarkSilkAddress(rec->address).IsValid())
            {
                CTxDestination address = CDarkSilkAddress(rec->address).Get();
                if (wallet->mapAddressBook.count(address))
                {
                    strHTML += "<b>" + tr("From") + ":</b> " + tr("unknown") + "<br>";
                    strHTML += "<b>" + tr("To") + ":</b> ";
                    strHTML += GUIUtil::HtmlEscape(rec->address);
                    if (!wallet->mapAddressBook[address].empty())
                        strHTML += " (" + tr("own address") + ", " + tr("label") + ": " + GUIUtil::HtmlEscape(wallet->mapAddressBook[address]) + ")";
                    else
                        strHTML += " (" + tr("own address") + ")";
                    strHTML += "<br>";
                }
            }
        }
    }

    //
    // To
    //
    if (wtx.mapValue.count("to") && !wtx.mapValue["to"].empty())
    {
        // Online transaction
        std::string strAddress = wtx.mapValue["to"];
        strHTML += "<b>" + tr("To") + ":</b> ";
        CTxDestination dest = CDarkSilkAddress(strAddress).Get();
        if (wallet->mapAddressBook.count(dest) && !wallet->mapAddressBook[dest].empty())
            strHTML += GUIUtil::HtmlEscape(wallet->mapAddressBook[dest]) + " ";
        strHTML += GUIUtil::HtmlEscape(strAddress) + "<br>";
    }

    //
    // Amount
    //
    if (wtx.IsCoinBase() && nCredit == 0)
    {
        //
        // Coinbase
        //
        CAmount nUnmatured = 0;
        BOOST_FOREACH(const CTxOut& txout, wtx.vout)
            nUnmatured += wallet->GetCredit(txout, ISMINE_ALL);
        strHTML += "<b>" + tr("Credit") + ":</b> ";
        if (wtx.IsInMainChain())
            strHTML += DarkSilkUnits::formatWithUnit(unit, nUnmatured)+ " (" + tr("matures in %n more block(s)", "", wtx.GetBlocksToMaturity()) + ")";
        else
            strHTML += "(" + tr("not accepted") + ")";
        strHTML += "<br>";
    }
    else if (nNet > 0)
    {
        //
        // Credit
        //
        strHTML += "<b>" + tr("Credit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, nNet) + "<br>";
    }
    else
    {
        bool fAllFromMe = true;
        BOOST_FOREACH(const CTxIn& txin, wtx.vin)
            fAllFromMe = fAllFromMe && wallet->IsMine(txin);

        bool fAllToMe = true;
        BOOST_FOREACH(const CTxOut& txout, wtx.vout)
            fAllToMe = fAllToMe && wallet->IsMine(txout);

        if (fAllFromMe)
        {
            //
            // Debit
            //
            BOOST_FOREACH(const CTxOut& txout, wtx.vout)
            {
                if (wallet->IsMine(txout))
                    continue;

                if (!wtx.mapValue.count("to") || wtx.mapValue["to"].empty())
                {
                    // Offline transaction
                    CTxDestination address;
                    if (ExtractDestination(txout.scriptPubKey, address))
                    {
                        strHTML += "<b>" + tr("To") + ":</b> ";
                        if (wallet->mapAddressBook.count(address) && !wallet->mapAddressBook[address].empty())
                            strHTML += GUIUtil::HtmlEscape(wallet->mapAddressBook[address]) + " ";
                        strHTML += GUIUtil::HtmlEscape(CDarkSilkAddress(address).ToString());
                        strHTML += "<br>";
                    }
                }

                strHTML += "<b>" + tr("Debit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, -txout.nValue) + "<br>";
            }

            if (fAllToMe)
            {
                // Payment to self
                CAmount nChange = wtx.GetChange();
                CAmount nValue = nCredit - nChange;
                strHTML += "<b>" + tr("Debit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, -nValue) + "<br>";
                strHTML += "<b>" + tr("Credit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, nValue) + "<br>";
            }
            CTransactionPoS txPoS;
            CTransaction tx = CTransaction(wtx);
            CAmount nTxFee = nDebit - txPoS.GetValueOut(tx);
            if (nTxFee > 0)
                strHTML += "<b>" + tr("Transaction fee") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, -nTxFee) + "<br>";
        }
        else
        {
            //
            // Mixed debit transaction
            //
            BOOST_FOREACH(const CTxIn& txin, wtx.vin)
                if (wallet->IsMine(txin))
                    strHTML += "<b>" + tr("Debit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, -wallet->GetDebit(txin, ISMINE_ALL)) + "<br>";
            BOOST_FOREACH(const CTxOut& txout, wtx.vout)
                if (wallet->IsMine(txout))
                    strHTML += "<b>" + tr("Credit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, wallet->GetCredit(txout, ISMINE_ALL)) + "<br>";
        }
    }

    strHTML += "<b>" + tr("Net amount") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, nNet, true) + "<br>";

    //
    // Message
    //
    if (wtx.mapValue.count("message") && !wtx.mapValue["message"].empty())
        strHTML += "<br><b>" + tr("Message") + ":</b><br>" + GUIUtil::HtmlEscape(wtx.mapValue["message"], true) + "<br>";
    if (wtx.mapValue.count("comment") && !wtx.mapValue["comment"].empty())
        strHTML += "<br><b>" + tr("Comment") + ":</b><br>" + GUIUtil::HtmlEscape(wtx.mapValue["comment"], true) + "<br>";

    strHTML += "<b>" + tr("Transaction ID") + ":</b> " + TransactionRecord::formatSubTxId(wtx.GetHash(), rec->idx) + "<br>";

    if (wtx.IsCoinBase() || wtx.IsCoinStake())
        strHTML += "<br>" + tr("DRKSLK from standard transactions confirm after 10 blocks and are spendable after just 1 block. Mined coins (PoW or PoS generated) must mature 100 blocks and be held for a minimum of 1 day before they can be staked or spent. When you generate a PoW or PoS block it is broadcast to the network to be added to the DarkSilk blockchain. If it fails to get into the chain, its state will change to \"not accepted\" and it won&apos;t be spendable. This may occasionally happen if another wallet/node generates a block within a few seconds of your wallet/node.") + "<br>";

    //
    // Debug view
    //
    if (fDebug)
    {
        strHTML += "<hr><br>" + tr("Debug information") + "<br><br>";
        BOOST_FOREACH(const CTxIn& txin, wtx.vin)
            if(wallet->IsMine(txin))
                strHTML += "<b>" + tr("Debit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, -wallet->GetDebit(txin, ISMINE_ALL)) + "<br>";
        BOOST_FOREACH(const CTxOut& txout, wtx.vout)
            if(wallet->IsMine(txout))
                strHTML += "<b>" + tr("Credit") + ":</b> " + DarkSilkUnits::formatWithUnit(unit, wallet->GetCredit(txout, ISMINE_ALL)) + "<br>";

        strHTML += "<br><b>" + tr("Transaction") + ":</b><br>";
        strHTML += GUIUtil::HtmlEscape(wtx.ToString(), true);

        CTxDB txdb("r"); // To fetch source txouts

        strHTML += "<br><b>" + tr("Inputs") + ":</b>";
        strHTML += "<ul>";

        BOOST_FOREACH(const CTxIn& txin, wtx.vin)
        {
            COutPoint prevout = txin.prevout;

            CTransaction prev;
            if(txdb.ReadDiskTx(prevout.hash, prev))
            {
                if (prevout.n < prev.vout.size())
                {
                    strHTML += "<li>";
                    const CTxOut &vout = prev.vout[prevout.n];
                    CTxDestination address;
                    if (ExtractDestination(vout.scriptPubKey, address))
                    {
                        if (wallet->mapAddressBook.count(address) && !wallet->mapAddressBook[address].empty())
                            strHTML += GUIUtil::HtmlEscape(wallet->mapAddressBook[address]) + " ";
                        strHTML += QString::fromStdString(CDarkSilkAddress(address).ToString());
                    }
                    strHTML = strHTML + " " + tr("Amount") + "=" + DarkSilkUnits::formatWithUnit(unit, vout.nValue);
                    strHTML = strHTML + " IsMine=" + (wallet->IsMine(vout) ? tr("true") : tr("false")) + "</li>";
                }
            }
        }

        strHTML += "</ul>";
    }

    strHTML += "</font></html>";
    return strHTML;
}
